### DISCUSSION TITLE
Layout Questions and Feedback

### DISCUSSION THREAD

**Discussion by zicklag** [emilk/egui#168] [Q&A]
Hey there, I just tried EGUI for the first time today and got my baisc GUI project up and going really quickly. It's pretty great. :smiley: 

So far the widgets seem pretty effective and you've got a lot of good features, but the one area I felt was a little lacking was the layout. I don't say this to critisize or anything, so don't take this the wrong way. Layout just felt a little cumbersome and like it wasn't easily possible to make responsive UI's that handle different screen sizes well.

I haven't gotten deep into it yet, and I think there are ways to make it do stuff I haven't found yet, but things like easily adding paddings or margins to frames with options for resizing or wrapping colums would be nice. And maybe all that I need is some basic examples to show how to do certain things. This is all my less-than-24-hours quick evaluation as I throw together a project so again, I'm not doing a heavy criticism of it. :slightly_smiling_face: 

That's just my first impression. I am really enjoying using this so far, and I think you've done well to deliver "The easiest to use GUI libary" ( not that I've tried many yet, but you're doing great! )

---

Also I notice you had this as a non-goal in the README: 

> Advanced and flexible layouts (that's fundamentally incompatible with immediate mode)

I just wanted to ask to get a better understanding of this myself, but something like using [stretch](https://github.com/vislyhq/stretch) for flexbox layout would be impossible, then correct? Is it because advanced layout systems takes too much time to run every frame? I was curious, even if it were a non-goal for EGUI if it would be possible to create an integration with `stretch` in another crate so I wanted to get a little more elaboration on "fundamentally incompatible" if you can spare the time, for my own education. :wink:

**Comment by zicklag** [comment]
Ah, I did find out a nicer way to do `Frame`'s with margins than I had found earlier:

```rust
Frame {
    margin: Vec2::new(6.0, 6.0),
    ..Default::default()
}
.show(ui, |ui| {
    // Put stuff in it
});
```

That's nicer than what I did earlier:

```rust
let mut frame = Frame::default();
frame.margin = Vec2::new(6.0, 6.0);
frame.show(ui, |ui| {
    // put stuff in it
});
```

But I'm thinking maybe we should still have a margin convenience function for doing that.

_I think I'll just use this thread to document my findings as I experiment. ;)_

**Reply by emilk** [reply]
Suggestions welcome, as are user stories (i.e. "this is why I want to do this").

> I think I'll just use this thread to document my findings as I experiment.

Sounds good!

**Comment by emilk** [comment]
Hi! Thanks for trying egui, and glad you like it :)

Yes, egui is still lacking when it comes to layouts. For instance, grid layouts was only added last month. Even when things are possible they are not well documented, and a big reason for this is because `egui` is still evolving quickly, and any docs will become outdated fast. The best chance is finding something similar to what you want [in the demo](https://emilk.github.io/egui/#demo) and try to copy that (see for instance the "Layout Test" window).

> I wanted to get a little more elaboration on "fundamentally incompatible" if you can spare the time, for my own education. 

Great question! In fact, I decided to spend some time to explain this in detail in the project `README.md`: <https://github.com/emilk/egui#why-immediate-mode>. As you can see, the fundamental problem is that most advanced layouts requires knowing the sizes of things before they can be positioned, but `egui` needs to know the positions of things before it can do the layout. You can still do nice layouts for fixed-sized widgets. For instance, if you have a bunch of images of known sizes you can probably use `stretch` to layout them and then just explicitly place them inside an `egui` `Ui` (https://docs.rs/egui/0.9.0/egui/struct.Ui.html#method.put)

**Reply by zicklag** [reply]
Awesome, that helps a lot, thanks! Good explanation in the readme.

I've had some fun with `egui` in the last few days and gotten a simple app working. I think my favorite part about it so far is the fact that it works so seamlessly on desktop *and* web. Your getting started template repo was also a huge timesaver. It let me write my custom calculator-type app in Rust and run it in the browser with no problems.

Anyway, I'll be keeping an eye on `egui`. Thanks for the lib!

**Reply by emilk** [reply]
That's great to hear, thanks for writing that! 😄
